File: java/src/sort/selection/SelectionSort.java
package sort.selection;

import java.util.ArrayList;
import java.util.List;

public class SelectionSort {
    public static List<Integer> sort(List<Integer> array) {
        List<Integer> result = new ArrayList<>();

        int size = array.size();
        for (int i = 0; i < size; i++) {
            int index = findSmallestIndex(array);
            result.add(array.get(index));
            array.remove(index);
        }

        return result;
    }

    /**
     * 查找最小元素位置索引。
     *
     * @param array
     * @return index 返回 -1 表示查找失败，否则表示元素在数组中的位置索引。
     */
    private static int findSmallestIndex(List<Integer> array) {
        int smallest = array.get(0), index = 0;
        for (int i = 0; i < array.size(); i++) {
            if (array.get(i) < smallest) {
                smallest = array.get(i);
                index = i;
            }
        }

        return index;
    }
}

File: php/sort/selection-sort/SelectionSort.php
<?php

class SelectionSort
{
    /**
     * 选择排序。
     *
     * @param array $array 待排序的数组。
     *
     * @return array
     */
    public static function sort(array $array): array
    {
        $result = [];
        foreach ($array as $value) {
            $index = self::findSmallestIndex($array);
            $result[] = array_splice($array, $index, 1)[0];
        }

        return $result;
    }

    /**
     * 查找最小值位置。
     *
     * @param array $array 待排序的数组
     *
     * @return int
     */
    private static function findSmallestIndex(array & $array): int
    {
        [$smallest, $index] = [$array[0], 0];

        foreach ($array as $key => $value) {
            if ($smallest > $value) {
                $smallest = $value;
                $index = $key;
            }
        }

        return $index;
    }
}
